support-figma/extended-layout-plugin/src/localization-module.ts

```diff
@@ -0,0 +1,274 @@
+  }
+  snakeCase = snakeCase.replace(/_{2,}/g, "_");
+  if (snakeCase.startsWith("_")) {
+    snakeCase = snakeCase.substring(1);
+  }
+  if (snakeCase.endsWith("_")) {
+    return snakeCase.substring(0, snakeCase.length - 1);
+  }
+  return snakeCase;
+}
+
+function isAllCaps(characters: string): boolean {
+  return characters.toUpperCase() === characters;
+}
+
+function countWords(characters: string): number {
+  // Remove leading/trailing whitespace and split into words
+  const words = characters.trim().split(/\W+/);
+
+  // Filter out empty strings (in case of multiple spaces)
+  return words.filter((word) => word !== "").length;
+}
+
+function saveToPluginData(node: TextNode, stringResName: string) {
+  node.setSharedPluginData(
+    PluginUtils.SHARED_PLUGIN_NAMESPACE,
+    STRING_RES_PLUGIN_DATA_KEY,
+    stringResName
+  );
+  console.log(
+    `### Save string res name ${node.name}-${node.id} with ${stringResName}`
+  );
+}
+
+function endsWithNumbers(stringResName: string): boolean {
+  var pattern = new RegExp("_\\d+$");
+  return pattern.test(stringResName);
+}
+
+function textMatches(
+  strRes: StringResource,
+  textRuns: string | string[]
+): boolean {
+  if (typeof strRes.text === typeof textRuns) {
+    if (typeof strRes.text === "string") {
+      return strRes.text === textRuns;
+    } else {
+      return (
+        strRes.text.length === textRuns.length &&
+        strRes.text.every((val, index) => val === textRuns[index])
+      );
+    }
+  }
+  return false;
+}
+
+function normalizeTextNode(node: TextNode): string | string[] {
+  console.log(`### normalize node ${node.id}`);
+  let stringArray: string[] = [];
+
+  // All styles.
+  try {
+    let segments = node.getStyledTextSegments([
+      "fontSize",
+      "fontWeight",
+      "letterSpacing",
+      "lineHeight",
+      "fontName",
+      "textDecoration",
+      "textCase",
+      "fills",
+      "textStyleId",
+      "fillStyleId",
+      // Comment out the following due to "Error: in getStyledTextSegments: Unknown list option"
+      // "listOptions",
+      "indentation",
+      "hyperlink",
+      "openTypeFeatures",
+      "boundVariables",
+    ]);
+    segments.forEach((it) => stringArray.push(normalizeString(it.characters)));
+  } catch (error) {
+    console.error("### An error occurred:", error);
+  }
+  if (stringArray.length > 1) {
+    return stringArray;
+  } else {
+    return normalizeString(node.characters);
+  }
+}
+
+function normalizeString(characters: string): string {
+  return characters
+    .replace(/(\r\n|\r|\n)/g, "\\n")
+    .replace(/\t/g, "\\t")
+    .replace(/\u2028/g, "\\u2028")
+    .replace(/&/g, "\\u0026")
+    .replace(/'/g, "\\'")
+    .replace(/"/g, '\\"')
+    .replace(/</g, "\\u003c")
+    .replace(/>/g, "\\u003e")
+    // preserve whitespaces at the beginning and the end
+    .replace(/^\s+/g, match => match.replace(/\s/g, "\\u0020"))
+    .replace(/\s+$/g, match => match.replace(/\s/g, "\\u0020"))
+    .normalize();
```

Comment:
Why is all this normalizing needed? Can you add some comments on why?

Reply:
Done.